### PR TITLE
Adds new max_inner_product vector similarity function

### PR DESCRIPTION
**Pull request by benwtrent**
Adds new `max_inner_product` vector similarity function. This differs from `dot_product` in the following ways:

 - Doesn't require vectors to be normalized
 - Scales the similarity between vectors differently to prevent negative scores

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 1]
```diff
@@ -662,6 +662,20 @@ float score(float similarity, ElementType elementType, int dim) {
...
+        },
+        MAX_INNER_PRODUCT(VectorSimilarityFunction.MAXIMUM_INNER_PRODUCT) {
+            @Override
+            float score(float similarity, ElementType elementType, int dim) {
+                return switch (elementType) {
+                    case BYTE, FLOAT -> {
+                        if (similarity < 1) {
+                            yield -1 * (1f / similarity - 1);
+                        } else {
+                            yield similarity - 1f;
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:674]: should this be `similarity + 1f`?

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 2]
```diff
@@ -662,6 +662,20 @@ float score(float similarity, ElementType elementType, int dim) {
...
                 };
             }
+        },
+        MAX_INNER_PRODUCT(VectorSimilarityFunction.MAXIMUM_INNER_PRODUCT) {
+            @Override
+            float score(float similarity, ElementType elementType, int dim) {
+                return switch (elementType) {
+                    case BYTE, FLOAT -> {
+                        if (similarity < 1) {
+                            yield -1 * (1f / similarity - 1);
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:672]: why this formula? why not `1 / (1 + -1 * similarity)`?

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java (hunk 1 above)
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:674]: no, this is transforming the score back to the actual maximum inner product. So it is the inverse of the mip transformation in Lucene:
```
  public static float scaleMaxInnerProductScore(float vectorDotProductSimilarity) {
    if (vectorDotProductSimilarity < 0) {
      return 1 / (1 + -1 * vectorDotProductSimilarity);
    }
    return vectorDotProductSimilarity + 1;
  }
```

Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:674]: @benwtrent This is not how I interpreted this `score` function; my interpretation is that given a similarity it produces a score based from similarity. This is also in line how it is implemented for other similarity?

Are things different for maxInnerProduct?

**Review by mayya-sharipova — APPROVED** [review]
@benwtrent Thanks Ben, LGTM

File: docs/changelog/99527.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+pr: 99445
```
Comment (otrosien) [inline docs/changelog/99527.yaml:1]: should be 99527 - this creates a wrong PR reference link in the release notes.

**Comment by benwtrent** [comment]
Thanks for pointing that out @otrosien ! I will see if we can fix that in the release notes.

File: docs/changelog/99527.yaml (hunk 3 above)
Comment (benwtrent) [reply docs/changelog/99527.yaml:1]: Fixing, https://github.com/elastic/elasticsearch/pull/101908 :D

**Pull request merged**